src/events.ts

```diff
@@ -0,0 +1,272 @@
+  ) {
+    return Promise.reject(new Error(spLog.timingReq));
+  }
+
+  return Promise.resolve(true);
+}
+
+/**
+ * Validates a consent-granted event
+ *
+ * @param argmap {Object} - the object to validate
+ * @returns - boolean promise
+ */
+function validateConsentGranted(argmap: ConsentGrantedProps): Promise<boolean> {
+  // validate type
+  if (!isObject(argmap)) {
+    return Promise.reject(new Error(spLog.evType));
+  }
+  // validate required props
+  if (
+    typeof argmap.expiry !== 'string'
+      || typeof argmap.documentId !== 'string'
+      || typeof argmap.version !== 'string'
+  ) {
+    return Promise.reject(new Error(spLog.consentGReq));
+  }
+
+  return Promise.resolve(true);
+}
+
+/**
+ * Validates a consent-withdrawn event
+ *
+ * @param argmap {Object} - the object to validate
+ * @returns - boolean promise
+ */
+function validateConsentWithdrawn(argmap: ConsentWithdrawnProps): Promise<boolean> {
+  // validate type
+  if (!isObject(argmap)) {
+    return Promise.reject(new Error(spLog.evType));
+  }
+  // validate required props
+  if (
+    typeof argmap.all !== 'boolean'
+      || typeof argmap.documentId !== 'string'
+      || typeof argmap.version !== 'string'
+  ) {
+    return Promise.reject(new Error(spLog.consentWReq));
+  }
+
+  return Promise.resolve(true);
+}
+
+/**
+ * Validates whether an object is valid ecommerce-item
+ *
+ * @param item {Object} - the object to validate
+ * @returns - boolean
+ */
+function isValidEcomItem(item: EcommerceItem): boolean {
+  if (
+    isObject(item)
+      && typeof item.sku === 'string'
+      && typeof item.price === 'number'
+      && typeof item.quantity === 'number'
+  ) {
+    return true;
+  }
+  return false;
+}
+
+/**
+ * Validates an array of ecommerce-items
+ *
+ * @param items {Object} - the object to validate
+ * @returns - boolean promise
+ */
+function validItemsArg(items: EcommerceItem[]): boolean {
+  return Object.prototype.toString.call(items) === '[object Array]'
+    && items
+      .map((i) => isValidEcomItem(i))
+      .reduce((acc,curr) => acc !== false && curr, true);
+}
+
+/**
+ * Validates an ecommerce-transaction event
+ *
+ * @param argmap {Object} - the object to validate
+ * @returns - boolean promise
+ */
+function validateEcommerceTransaction(argmap: EcommerceTransactionProps): Promise<boolean> {
+  // validate type
+  if (!isObject(argmap)) {
+    return Promise.reject(new Error(spLog.evType));
+  }
+  // validate required props
+  if (
+    typeof argmap.orderId !== 'string'
+      || typeof argmap.totalValue !== 'number'
+      || !validItemsArg(argmap.items)
+  ) {
+    return Promise.reject(new Error(spLog.ecomReq));
+  }
+
+  return Promise.resolve(true);
+}
+
+export {
+  isValidSD,
+  validateContexts,
+  validateSelfDesc,
+  validateScreenView,
+  validateStructured,
+  validatePageView,
+  validateTiming,
+  validateConsentGranted,
+  validateConsentWithdrawn,
+  isValidEcomItem,
+  validItemsArg,
+  validateEcommerceTransaction,
```

Comment:
I think this is probably the only tracker that prevalidates events before sending them. I see some value in it but I'm still not 100% convinced. My main concern is that if we change an event definition (especially those that are SelfDescribing under the hood) then we need to remember to change this or else it won't work. Some of these calls aren't cheap either, iterating through the entire list of ecom items is a good example. The big question is should be do this in the tracker to catch (some) bad events earlier or should we defer this always to the pipeline to validate events? 
I remember us discussing this in a call, but I think it would be worth recapping the thoughts on why we might want to validate events in this tracker before passing the events to the iOS and Android trackers, whereas we don't do this in the "native" trackers.

Reply:
Indeed, i've been changing my mind on this. 
On one hand the checks are redundant, since there is some validation in the native trackers too. 
On the other hand, some parts of this validation offer some guarantees in the bridge, so that we don't have to, for example, check if arrays are arrays or have if-else statements even for the required event props in order to construct the event. 

For example, a screenViewEvent without `name` at the moment will not reach the bridge at all. But if we removed the check in JavaScript-side, then it would need to be done in the bridge-side (i'll have another look to confirm my assumption here). Would that be better from performance perspective? 
Or do you mean to just let it error in the bridge? That would be fine for the Android bridge, but i was not confident of error handling in ObjC, where try-catching somehow seemed to me hard to trust.

Then as you said, there is the bad events side of things. At the moment, there are only checks for the required props. And that is only for the javascript users that have no type-checking, or for a runtime edge case. Not much of an argument for checking indeed here.

Concerning possible changes to event definitions, would that be a breaking change in the native trackers?
Generally, is there a right balance? What do you think?

Reply:
I'm happy for this to stay in, it isn't too heavy and seems safer to have it than not given the brigding in this application. Even if a schema did update, you'd need to update the native tracker and therefore update this tracker too. Perhaps something we can look in to more in the future but these feels like a good compromise at the moment.